### PR TITLE
Update run_glue for do_predict with local test data (#9442)

### LINKED ISSUE #9442: [examples/text-classification] `do_predict` for the test set of local datasets
# 🚀 Feature request

It seems that `run_glue.py` has the train set and the validation set management for local CSV/JSON files, but it doesn't have args for managing the test set of the local datasets.

https://github.com/huggingface/transformers/blob/7a9f1b5c99e9a5d1772649d029acdf5160419239/examples/text-classification/run_glue.py#L90-L95

I think the script is intended to be used not only for the train/validation but the test, as `glue` tasks test sets are downloaded as shown in https://huggingface.co/docs/datasets/loading_datasets.html#selecting-a-configuration.
It has the `--do_predict` option for the test sets.

If there is no particular reason for not having the ability to read the test set in the local dataset, would it be ok for me to add the feature?
Or is there some intention behind this implementation?

## Motivation

I'd like to train, validate, and test my own local dataset.

## Your contribution

I think some modifications like the below may help to add the feature.

``` python
    test_file: Optional[str] = field(
        default=None, metadata={"help": "A csv or a json file containing the test data."}
    )
```

``` python
        datasets = load_dataset(
            "csv", data_files={"train": data_args.train_file, "validation": data_args.validation_file, "test": data_args.test_file}
        )
```

``` python
    # if data_args.task_name is not None:
    #     test_dataset = datasets["test_matched" if data_args.task_name == "mnli" else "test"]
    test_dataset = datasets["test_matched" if data_args.task_name == "mnli" else "test"]
```

Thank you in advance.

### PR DESCRIPTION
**Pull request by forest1988**
# What does this PR do?

Currently, run_glue.py cannot use the test set (`do_predict`) unless we give it a GLUE task name.
This PR will allow us to use the local test dataset.

As commented in #9442, I tried to achieve the functionality with only simple changes.

- It still works with only the local train and valid files (in other words, this PR does not break the current operation.).
- If we add `--do_predict` with out adding specific params, we will get an error statement saying that we need either the GLUE task name or the path of the local test file.

Fixes #9442 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@sgugger

Thank you for your kind comments on the issue. 
I have tried to keep it simple and hope there is no problem as an example script.

### CODE REVIEW

**Comment by forest1988** [comment]
Error message of the Circle CI is:

```
-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================== short test summary info ============================
FAILED tests/test_pipelines_conversational.py::SimpleConversationPipelineTests::test_history_cache
FAILED tests/test_pipelines_conversational.py::SimpleConversationPipelineTests::test_integration_torch_conversation
==== 2 failed, 4207 passed, 1744 skipped, 734 warnings in 190.84s (0:03:10) ====
```

```
FAILED tests/test_pipelines_conversational.py::SimpleConversationPipelineTests::test_history_cache
==== 1 failed, 4178 passed, 1774 skipped, 735 warnings in 260.31s (0:04:20) ====
```

I'm sorry but I'd like to ask you if `run_glue.py` is related to the conversation pipeline.

File: examples/text-classification/run_glue.py [hunk 1]
```diff
@@ -413,6 +414,25 @@ def compute_metrics(p: EvalPrediction):
...
+
+        if data_args.task_name is None and data_args.test_file is not None:
+            extension = data_args.test_file.split(".")[-1]
+            assert extension in ["csv", "json"], "`test_file` should be a csv or a json file."
+            if data_args.test_file.endswith(".csv"):
+                # Loading a dataset from a local csv file
+                test_dataset = load_dataset("csv", data_files={"test": data_args.test_file})
+            else:
+                # Loading a dataset from a local json file
+                test_dataset = load_dataset("json", data_files={"test": data_args.test_file})
```
Comment (sgugger) [inline examples/text-classification/run_glue.py:428]: Can we put those lines earlier, with the validation dataset? This way the map will be done with the other dataset. I think we can do something nice by creating `data_files={"train": data_args.train_file, "validation": data_args.validation_file}` and then adding the keys `test` if the test_file is passed.

**Review by sgugger — APPROVED** [review]
The failures seems spurious indeed. I've left some comment to move the test dataset creation a bit, but it overall looks good to me.

File: examples/text-classification/run_glue.py (hunk 1 above)
Comment (forest1988) [reply examples/text-classification/run_glue.py]: Thank you for your comment! I've reflected the review.
The nested `if` statements in the code have increased, but I think the readability may have improved in terms of "whether to use GLUE task or to use local files". What do you think?

Comment (forest1988) [reply examples/text-classification/run_glue.py]: I also added the logger output to make sure that the local files a user wants to use are loaded correctly. If this is superfluous, please let me know and I will remove it.

File: examples/text-classification/run_glue.py [hunk 2]
```diff
@@ -206,16 +206,34 @@ def main():
...
+        # Loading a dataset from your local files.
+        # CSV/JSON training and evaluation files are needed.
+        data_files = {"train": data_args.train_file, "validation": data_args.validation_file}
+
+        # Get the test dataset: you can provide your own CSV/JSON test file (see below)
+        # when you use `do_predict` without specifying a GLUE benchmark task.
+        if training_args.do_predict:
+            if data_args.test_file is not None:
+                extension = data_args.test_file.split(".")[-1]
+                assert extension in ["csv", "json"], "`test_file` should be a csv or a json file."
```
Comment (sgugger) [inline examples/text-classification/run_glue.py:219]: The extension will need to be the same one as for the training and validation file, so we should adapt this assert to test that.

File: examples/text-classification/run_glue.py [hunk 3]
```diff
@@ -206,16 +206,34 @@ def main():
...
+        if data_args.train_file.endswith(".csv"):
+            # Loading a dataset from local csv files
+            datasets = load_dataset(
+                "csv", data_files=data_files
+            )
+        else:
+            # Loading a dataset from local json files
+            datasets = load_dataset(
+                "json", data_files=data_files
+            )
```
Comment (sgugger) [inline examples/text-classification/run_glue.py]:
```suggestion
            datasets = load_dataset("json", data_files=data_files)
```
Can fit in one line now :-)

**Review by sgugger — APPROVED** [review]
Looking good! Left two more small comments and it should be good to merge!

File: examples/text-classification/run_glue.py [hunk 4]
```diff
@@ -205,16 +206,30 @@ def main():
...
+        if training_args.do_predict:
+            if data_args.test_file is not None:
+                extension = data_args.test_file.split(".")[-1]
+                assert extension in ["csv", "json"], "`test_file` should be a csv or a json file."
+                data_files["test"] = data_args.test_file
+            else:
+                raise ValueError("Need either a GLUE task or a test file for `do_predict`.")
+
+        for key in data_files.keys():
+            logger.info(f"load a local file for {key}: {data_files[key]}")
```
Comment (sgugger) [inline examples/text-classification/run_glue.py:225]: This could info to log, thanks for adding!

File: examples/text-classification/run_glue.py (hunk 3 above)
Comment (forest1988) [reply examples/text-classification/run_glue.py]: Thank you!　
It may be that the old code before applying the auto-format was left in place. 
I have applied the auto-format in b2936c3, could you please check if it is fit in one line?

File: examples/text-classification/run_glue.py (hunk 2 above)
Comment (forest1988) [reply examples/text-classification/run_glue.py]: Reflecting the comments, assert now checks that the `test file` has the same extension as the `train file`.
Also, I thought there was no check if the `validation file` has the same extension as the `train file`, so I modified that. Is this change OK?

**Review by LysandreJik — APPROVED** [review]
Great, thanks for adding!

**Comment by forest1988** [comment]
@sgugger @LysandreJik 
Thank you for reviewing and merging!
